overlord/devicestate/handlers.go

```diff
@@ -114,12 +130,33 @@ func (m *DeviceManager) doSetModel(t *state.Task, _ *tomb.Tomb) error {
 		if snapst.Flags.Required && !requiredSnaps.Contains(naming.Snap(snapName)) {
 			snapst.Flags.Required = false
 			snapstate.Set(st, snapName, snapst)
+			cleanedRequiredSnaps = append(cleanedRequiredSnaps, snapName)
 		}
 		// TODO: clean "required" flag of "core" if a remodel
 		//       moves from the "core" snap to a different
 		//       bootable base snap.
 	}
 
+	// ensure  we undo the cleanedRequiredSnaps if e.g. remodCtx
+	defer func() {
+		if err == nil {
```

Comment:
This kind of code is only working if there's one definition of `err` in the function.
Anything that accidentally locally defines error will silently compile and misbehave.

I wonder what we could do to make it more robust. Perhaps a protocol where a private function takes a `failureRecovery` interface / type and calls functions that effectively record operations to be undone in case of failure. The public function creates the failure recovery helper and calls the private function, invoking recovery if it fails.

Reply:
discussed on what this relies on IRC:

> <pedronis> zyga: is the this bit of the spec: A "return" statement that specifies results sets the result parameters before any deferred functions are executed.
> <pedronis> https://golang.org/ref/spec#Return_statement